temporal-sdk/src/main/java/io/temporal/internal/sync/SignalDispatcher.java

```diff
@@ -0,0 +1,162 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.sync;
+
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.common.converter.DataConverter;
+import io.temporal.common.converter.DataConverterException;
+import io.temporal.common.converter.EncodedValues;
+import io.temporal.common.interceptors.WorkflowInboundCallsInterceptor;
+import io.temporal.common.interceptors.WorkflowOutboundCallsInterceptor;
+import io.temporal.internal.metrics.MetricsType;
+import io.temporal.workflow.DynamicSignalHandler;
+import io.temporal.workflow.Workflow;
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Queue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class SignalDispatcher {
+
+  private static class SignalData {
+    private final String signalName;
+    private final Optional<Payloads> payload;
+    private final long eventId;
+
+    private SignalData(String signalName, Optional<Payloads> payload, long eventId) {
+      this.signalName = Objects.requireNonNull(signalName);
+      this.payload = Objects.requireNonNull(payload);
+      this.eventId = eventId;
+    }
+
+    public String getSignalName() {
+      return signalName;
+    }
+
+    public Optional<Payloads> getPayload() {
+      return payload;
+    }
+
+    public long getEventId() {
+      return eventId;
+    }
+  }
+
+  private static final Logger log = LoggerFactory.getLogger(SignalDispatcher.class);
+
+  private WorkflowInboundCallsInterceptor inboundCallsInterceptor;
+  private final DataConverter converter;
+
+  private final Map<String, WorkflowOutboundCallsInterceptor.SignalRegistrationRequest>
+      signalCallbacks = new HashMap<>();
+
+  private DynamicSignalHandler dynamicSignalHandler;
+
+  /** Buffers signals which don't have a registered listener. */
+  private final Queue<SignalData> signalBuffer = new ArrayDeque<>();
```

Comment:
Does this need a size limit?

Reply:
The history sent by the service has a hard limit on a number of signals it can contain.